README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

ikChatBot Flutter Package

The ikChatBot Flutter package provides an easy way to integrate a customizable chatbot into your Flutter applications. This package allows you to quickly add a chat interface with interactive responses and user-friendly configuration options. With ikChatBot, you can enhance user engagement and provide automated assistance through a chat-based interface.


## Features
![thumbnail](https://github.com/iksoft/ikchatbot/assets/62053085/bdf6530e-475a-424f-a065-b60cfe8bcdc7)

[//]: # (<div align="center">)

[//]: # (    <img src="https://github.com/iksoft/ikchatbot/assets/62053085/3651c184-d70b-432a-a387-25a87b7128cd" alt="Screenshot 1" width="200"/>)

[//]: # (    <img src="https://github.com/iksoft/ikchatbot/assets/62053085/62039c04-c9ec-4264-b9ef-97d2aac7d502" alt="Screenshot 2" width="200"/>)

[//]: # (    <img src="https://github.com/iksoft/ikchatbot/assets/62053085/8d675799-4fe0-4ff0-b6f6-7dadf64e0502" alt="Screenshot 2" width="200"/>)

[//]: # (    <img src="https://github.com/iksoft/ikchatbot/assets/62053085/6a598e63-091d-433e-982e-42cb8561c9ee" alt="Screenshot 2" width="200"/>)

[//]: # (   )
[//]: # (</div>)


## Getting started

# whatsapp_chatbot Package

The `whatsapp_chatbot` package provides a customizable chatbot widget for Flutter applications. This widget allows you to easily integrate a chatbot into your app with various customization options.

## Features

- Customizable chatbot appearance
- Customizable chat bubble colors
- Initial greeting and default responses
- Placeholder for user input
- Easy integration into Flutter apps
- Use Bot typing.. features
- Make direct calls
- Chat directly on whatsapp


Whether you're building a customer support app or just looking to add some interactivity to your application, ikChatBot can help you create a dynamic chatbot experience tailored to your needs.

## Getting Started

To use this package, add `whatsapp_chatbot` to your `pubspec.yaml` file:

```yaml
dependencies:
  ikchatbot: ^0.0.1
```
Import the package to your screen
```yaml
import 'package:whatsapp_chatbot/whatsapp_chatbot.dart';
```
Copy and Paste this function to use the features of the plugin
```yaml
final config = Config(
botDelay: 3,
waitText: 'Bot Thinking...',
defaultResponseMessage: "Sorry! I didn't catch that!\nPlease try again!",
keywords: ['hello', 'hi', 'how are you',],
response: [
  'Hi\nHow can I assist you today?',
  'Hello!\nHow can I be of help?',
  'I am doing great!',

],
greetings: "Hi there👋🏾\nHow can I help you?",
headerText: 'Iksoft Technologies',
subHeaderText: 'Online',
profileImage: 'profileImage',
buttonText: 'Start Chat',
buttonColor: const Color.fromARGB(255, 73, 4, 4),
chatIcon: const Icon(Icons.person),
headerColor: const Color.fromARGB(255, 73, 4, 4),
message: 'Hello! This is a direct WhatsApp message.',
phoneNumber: '+233550138086',
chatBackgroundColor: const Color.fromARGB(255, 238, 231, 223),
onlineIndicator: const Color.fromARGB(255, 37, 211, 102),
  );
```

Full Example on how to use this plugin
```yaml
import 'package:flutter/material.dart';
  import 'package:whatsapp_chatbot/whatsapp_chatbot.dart';

  void main() {
  runApp(const MyApp());
}

  class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
  return MaterialApp(
title: 'Flutter Demo',
theme: ThemeData(
  colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
  useMaterial3: true,
  ),
home: HomePage(),
  );
}
}

  class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

  class _HomePageState extends State<HomePage> {
  final List<String> keywords = [];

  final config = Config(
botDelay: 3,
waitText: 'Bot Thinking...',
defaultResponseMessage: "Sorry! I didn't catch that!\nPlease try again!",
keywords: ['hello', 'hi', 'how are you',],
response: [
  'Hi\nHow can I assist you today?',
  'Hello!\nHow can I be of help?',
  'I am doing great!',

],
greetings: "Hi there👋🏾\nHow can I help you?",
headerText: 'Iksoft Technologies',
subHeaderText: 'Online',
profileImage: 'profileImage',
buttonText: 'Start Chat',
buttonColor: const Color.fromARGB(255, 73, 4, 4),
chatIcon: const Icon(Icons.person),
headerColor: const Color.fromARGB(255, 73, 4, 4),
message: 'Hello! This is a direct WhatsApp message.',
phoneNumber: '+233550138086',
chatBackgroundColor: const Color.fromARGB(255, 238, 231, 223),
onlineIndicator: const Color.fromARGB(255, 37, 211, 102),
  );

  @override
  Widget build(BuildContext context) {
  return Scaffold(

body: Stack(
  children: [
    WhatsappChatBot(
    settings: config,
    )
  ],
  ));
}
}

```

## Additional information

To find more information, check out the package documentation.

Contributions are welcome! If you'd like to contribute to this package, please submit issues or pull requests on GitHub.


If you encounter any issues or need support, feel free to file issues on GitHub or contact us at aduamahdaniel16@gmail.com.

Our team will do their best to respond promptly to issues and inquiries.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.0.1] - YYYY-MM-DD
### Added
- Initial release of `whatsapp_chatbot` package.
- Customizable chatbot appearance
- Customizable chat bubble colors
- Initial greeting and default responses
- Placeholder for user input
- Easy integration into Flutter apps
- Use Bot typing.. features
- Make direct calls
- Chat directly on whatsapp


[Unreleased]: https://github.com/iksoft/ikchatbot/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/iksoft/ikchatbot/releases/tag/v1.0.0

lib/src/popup.dart
import 'package:flutter/material.dart';
import 'package:font_awesome_flutter/font_awesome_flutter.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:whatsapp_chatbot/src/chat_bubble.dart';
import 'package:whatsapp_chatbot/src/config.dart';

class WhatsappChatBot extends StatefulWidget {
  final Config settings;

  const WhatsappChatBot({Key? key, required this.settings}) : super(key: key);

  @override
  State<WhatsappChatBot> createState() => _WhatsappChatBotState();
}

class _WhatsappChatBotState extends State<WhatsappChatBot> {
  late TextEditingController userText = TextEditingController();
  late bool isVisible = false;
  late bool isChat = true;
  late bool isChat2 = false;
  late bool _isWaitingForUserResponse = true;

  final List<ChatMessage> _message = [];

  final ScrollController _scrollController = ScrollController();

  @override
  void initState() {
    super.initState();
    _message.add(ChatMessage(
      isUser: false,
      messageText: widget.settings.greetings,
      messageTime: DateTime.now(),
    ));

    // Add buttons as a separate list of messages
    _addButtonsMessages();
  }

  void _sendMessage() async {
    final url =
        'https://wa.me/${widget.settings.phoneNumber}/?text=${Uri.encodeFull(widget.settings.message)}';

    try {
      if (await canLaunch(url)) {
        await launch(url);
      } else {
        throw 'Could not launch $url';
      }
    } catch (e) {
      print(e);
    }
  }

  void _addButtonsMessages() {
    // Add each button as a separate message in the list
    _message.add(ChatMessage(
      isUser: false,
      messageText: "Quick Actions",
      messageTime: DateTime.now(),
      additionalWidget: SizedBox(
        height: 50,
        width: 200,
        child: ListView(
          children: [
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 8.0),
              child: InkWell(
                onTap: () {
                  _sendMessage();
                },
                child: Container(
                  decoration: BoxDecoration(
                      color: widget.settings.buttonColor,
                      borderRadius: BorderRadius.circular(30.0)),
                  width: 100,
                  height: 50,
                  child: const Center(
                      child: Text(
                    "Live Chat",
                    style: TextStyle(color: Colors.white),
                  )),
                ),
              ),
            ),
          ],
        ),
      ),
    ));
  }

  void _startTypingAnimation(String reply) {
    final typingMessage = ChatMessage(
      isUser: false,
      messageText: widget.settings.waitText,
      messageTime: DateTime.now(),
    );

    setState(() {
      _message.add(typingMessage);
      widget.settings.subHeaderText = "Typing...";
    });

    Future.delayed(Duration(seconds: widget.settings.botDelay), () {
      _message.remove(typingMessage);
      widget.settings.subHeaderText = "Online";
      _addBotMessage(reply);
    });
  }

  void _addBotMessage(String text) {
    final botMessage = ChatMessage(
      isUser: false,
      messageText: text,
      messageTime: DateTime.now(),
    );

    setState(() {
      _message.add(botMessage);
    });
  }

  int _findMatchingKeyword(String response) {
    for (int i = 0; i < widget.settings.keywords.length; i++) {
      if (response.contains(widget.settings.keywords[i])) {
        return i;
      }
    }
    return -1;
  }

  void _handleSubmitted(String text) {
    _handleUserResponse(text);
  }

  void _handleUserResponse(String response) {
    final lowerCaseResponse = response.trim().toLowerCase();

    late String reply;
    late String userSelectedOption;
    final int index = _findMatchingKeyword(lowerCaseResponse);

    if (index != -1) {
      userSelectedOption = response;
      reply = widget.settings.response[index];
    } else {
      userSelectedOption = response;
      reply = widget.settings.defaultResponseMessage;
    }

    setState(() {
      _message.add(ChatMessage(
        isUser: true,
        messageText: userSelectedOption,
        messageTime: DateTime.now(),
      ));
      userText.clear();
    });
    if (reply.isNotEmpty) {
      _isWaitingForUserResponse = false;
      _startTypingAnimation(reply);
    }
  }

  void _addSuggestions(List<String> suggestions) {
    List<Widget> suggestionButtons = [];
    suggestions.forEach((suggestion) {
      suggestionButtons.add(
        ElevatedButton(
          onPressed: () {
            setState(() {
              _handleSubmitted(suggestion);
            });
          },
          child: Text(suggestion),
        ),
      );
    });

    setState(() {
      _message.add(
        ChatMessage(
          isUser: false,
          messageText: 'Please select an option:',
          messageTime: DateTime.now(),
        ),
      );
      _message.add(
        ChatMessage(
          isUser: false,
          messageText: '',
          messageTime: DateTime.now(),
          additionalWidget: Row(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: suggestionButtons,
          ),
        ),
      );
    });
  }

  void _makePhoneCall(String phoneNumber) async {
    String number = 'tel:$phoneNumber';
    var uri = Uri.parse(number);
    try {
      if (await canLaunchUrl(uri)) {
        await launch(uri.toString());
      } else {
        throw 'Could not Launch the call number $number';
      }
    } catch (e) {
      print(e);
    }
  }

  Widget build(BuildContext context) {
    return Stack(
      children: [
        // Widget positioned at the bottom right
        Positioned(
          bottom: 100,
          right: 20,
          child: Visibility(
            visible: isVisible,
            child: Container(
              decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.circular(10.0),
                boxShadow: [
                  BoxShadow(
                    color: Colors.grey.withOpacity(0.5),
                    spreadRadius: 3,
                    blurRadius: 7,
                    offset: const Offset(0, 3), // changes position of shadow
                  ),
                ],
              ),
              width: 350,
              height: 500,
              child: Column(
                children: [
                  Expanded(
                    flex: 1,
                    child: Container(
                      decoration: BoxDecoration(
                          color: widget.settings.headerColor,
                          borderRadius: const BorderRadius.only(
                              topLeft: Radius.circular(10),
                              topRight: Radius.circular(10))),
                      child: Center(
                        child: ListTile(
                          leading: Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: Icon(
                              FontAwesomeIcons.robot,
                              color: Colors.white,
                              size: 40,
                            ),
                           
                          ),
                          title: Text(
                            widget.settings.headerText,
                            style: const TextStyle(color: Colors.white),
                          ),
                          subtitle: Row(
                            children: [
                              Padding(
                                padding: const EdgeInsets.all(8.0),
                                child: Container(
                                  width: 12,
                                  height: 12,
                                  color: widget.settings.onlineIndicator,
                                ),
                              ),
                              Text(
                                widget.settings.subHeaderText,
                                style: const TextStyle(color: Colors.white),
                              ),
                            ],
                          ),
                          trailing: IconButton(
                            icon: const Icon(
                              Icons.call,
                              color: Colors.white,
                            ),
                            onPressed: () {
                              _makePhoneCall(widget.settings.phoneNumber);
                            },
                          ),
                        ),
                      ),
                    ),
                  ),
                  Expanded(
                      flex: 3,
                      child: Container(
                          color: widget.settings.chatBackgroundColor,
                          child: ListView.builder(
                            controller: _scrollController,
                            reverse: true,
                            itemCount: _message.length,
                            itemBuilder: (_, int index) => Padding(
                              padding: const EdgeInsets.all(8.0),
                              child: _message.reversed.toList()[index],
                            ),
                          ))),
                  Visibility(
                    visible: isChat,
                    child: Expanded(
                      flex: 1,
                      child: Container(
                        decoration: const BoxDecoration(
                            color: Colors.white,
                            borderRadius: BorderRadius.only(
                                bottomLeft: Radius.circular(10),
                                bottomRight: Radius.circular(10))),
                        child: SizedBox(
                          width: MediaQuery.of(context).size.width / 1,
                          child: Padding(
                            padding: const EdgeInsets.all(18.0),
                            child: FloatingActionButton.extended(
                                elevation: 0,
                                backgroundColor: widget.settings.buttonColor,
                                onPressed: () {
                                  setState(() {
                                    isChat = !isChat;
                                    isChat2 = !isChat2;
                                  });
                                },
                                label: Text(
                                  widget.settings.buttonText,
                                  style: const TextStyle(
                                      fontSize: 18.0,
                                      color: Colors.white,
                                      fontWeight: FontWeight.bold),
                                )),
                          ),
                        ),
                      ),
                    ),
                  ),
                  Visibility(
                    visible: isChat2,
                    child: Expanded(
                      flex: 1,
                      child: Container(
                        decoration: const BoxDecoration(
                            color: Colors.white,
                            borderRadius: BorderRadius.only(
                                bottomLeft: Radius.circular(10),
                                bottomRight: Radius.circular(10))),
                        child: SizedBox(
                          width: MediaQuery.of(context).size.width / 1,
                          child: Padding(
                              padding:
                                  const EdgeInsets.symmetric(horizontal: 18.0),
                              child: Column(
                                children: [
                                  TextField(
                                    onChanged: (text) {
                                      if (!_isWaitingForUserResponse) {
                                        _isWaitingForUserResponse = true;
                                      }
                                    },
                                    controller: userText,
                                    decoration: InputDecoration(
                                        hintText: 'Enter Message',
                                        suffixIcon: IconButton(
                                          icon: const Icon(Icons.send),
                                          onPressed: () {
                                            setState(() {
                                              _handleSubmitted(userText.text);
                                            });
                                          },
                                        ),
                                        border: InputBorder.none),
                                  ),
                                  const Center(
                                    child: Text(
                                      'Powered by afrilen.com',
                                      style: TextStyle(fontSize: 12),
                                    ),
                                  )
                                ],
                              )),
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
        // Widget for FloatingActionButton
        Positioned(
          bottom: 30,
          right: 20,
          child: FloatingActionButton(
            backgroundColor: isVisible
                ? const Color.fromARGB(255, 107, 3, 3)
                : const Color.fromARGB(255, 37, 211, 102),
            onPressed: () {
              setState(() {
                isVisible = !isVisible;
              });
            },
            child: isVisible
                ? const Icon(
                    Icons.close,
                    color: Colors.white,
                  )
                : const Icon(
                    FontAwesomeIcons.whatsapp,
                    color: Colors.white,
                  ),
          ),
        ),
      ],
    );
  }
}


lib/src/config.dart
import 'package:flutter/material.dart';

class Config {
  final List<String> keywords;
  final List<String> response;
  final greetings;
  final headerText;
  String subHeaderText;
  final profileImage;
  final buttonText;
  final chatIcon;
  final headerColor;
  final buttonColor;
  final phoneNumber;
  final message;
  final defaultResponseMessage;
  final waitText;
  final int botDelay;
  final onlineIndicator;
  final chatBackgroundColor;

  Config({
    required this.chatBackgroundColor,
    required this.onlineIndicator,
    required this.botDelay,
    required this.waitText,
    required this.defaultResponseMessage,
    required this.keywords,
    required this.response,
    required this.greetings,
    required this.headerText,
    required this.subHeaderText,
    required this.profileImage,
    required this.buttonText,
    required this.buttonColor,
    required this.chatIcon,
    required this.headerColor,
    required this.message,
    required this.phoneNumber,
  });
}


lib/src/chat_bubble.dart
import 'package:flutter/material.dart';
import 'package:whatsapp_chatbot/whatsapp_chatbot.dart';

class ChatMessage extends StatefulWidget {
  final Config? settings;
  final String messageText;
  final bool isUser;
  final DateTime messageTime;
  final Widget? additionalWidget; // Added additionalWidget

  ChatMessage({
    this.settings,
    required this.isUser,
    required this.messageText,
    required this.messageTime,
    this.additionalWidget, // Made additionalWidget nullable
  });

  @override
  State<ChatMessage> createState() => _ChatMessageState();
}

class _ChatMessageState extends State<ChatMessage> {
  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment:
          widget.isUser ? CrossAxisAlignment.end : CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.symmetric(vertical: 5.0),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            mainAxisAlignment:
                widget.isUser ? MainAxisAlignment.end : MainAxisAlignment.start,
            children: [
              if (!widget.isUser)
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: CircleAvatar(
                    backgroundColor: widget.settings?.headerColor,
                    radius: 20.0,
                    child: const Icon(Icons.person),
                  ),
                ),
              Column(
                children: [
                  Container(
                    constraints: BoxConstraints(
                      maxWidth: MediaQuery.of(context).size.width * 0.7,
                    ),
                    padding: const EdgeInsets.all(10.0),
                    decoration: BoxDecoration(
                      color: widget.isUser
                          ? const Color.fromARGB(255, 222, 246, 202)
                          : const Color.fromARGB(255, 252, 250, 250),
                      borderRadius: widget.isUser
                          ? const BorderRadius.only(
                              topLeft: Radius.circular(12.0),
                              bottomLeft: Radius.circular(12.0),
                              bottomRight: Radius.circular(12.0),
                            )
                          : const BorderRadius.only(
                              topRight: Radius.circular(12.0),
                              bottomLeft: Radius.circular(12.0),
                              bottomRight: Radius.circular(12.0),
                            ),
                    ),
                    child: Column(
                      crossAxisAlignment: widget.isUser
                          ? CrossAxisAlignment.end
                          : CrossAxisAlignment.start,
                      mainAxisAlignment: widget.isUser
                          ? MainAxisAlignment.end
                          : MainAxisAlignment.start,
                      children: [
                        Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 8.0),
                          child: Text(
                            widget.messageText,
                            style: widget.isUser
                                ? const TextStyle(color: Colors.black)
                                : const TextStyle(color: Colors.black),
                          ),
                        ),
                        Padding(
                          padding: const EdgeInsets.all(8.0),
                          child: Text(
                            "${widget.messageTime.hour}:${widget.messageTime.minute}:${widget.messageTime.second}",
                            style: const TextStyle(
                                color: Colors.grey, fontSize: 10.0),
                          ),
                        ),
                        if (widget.additionalWidget !=
                            null) // Display additional widget if available
                          Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: widget.additionalWidget,
                          ),
                      ],
                    ),
                  ),
                ],
              ),
              if (widget.isUser)
                const Padding(
                  padding: EdgeInsets.all(8.0),
                  child: CircleAvatar(
                    backgroundColor: Color.fromARGB(255, 7, 94, 84),
                    radius: 20.0,
                    child: Icon(
                      Icons.person,
                      color: Colors.white,
                    ),
                  ),
                ),
            ],
          ),
        )
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:whatsapp_chatbot/whatsapp_chatbot.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final List<String> keywords = [];

  final config = Config(
    botDelay: 3,
    waitText: 'Bot Thinking...',
    defaultResponseMessage: "Sorry! I didn't catch that!\nPlease try again!",
    keywords: ['hello', 'hi', 'how are you',],
    response: [
      'Hi\nHow can I assist you today?',
      'Hello!\nHow can I be of help?',
      'I am doing great!',

    ],
    greetings: "Hi there👋🏾\nHow can I help you?",
    headerText: 'Iksoft Technologies',
    subHeaderText: 'Online',
    profileImage: 'profileImage',
    buttonText: 'Start Chat',
    buttonColor: const Color.fromARGB(255, 73, 4, 4),
    chatIcon: const Icon(Icons.person),
    headerColor: const Color.fromARGB(255, 73, 4, 4),
    message: 'Hello! This is a direct WhatsApp message.',
    phoneNumber: '+233550138086',
    chatBackgroundColor: const Color.fromARGB(255, 238, 231, 223),
    onlineIndicator: const Color.fromARGB(255, 37, 211, 102),
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(

        body: Stack(
          children: [
            WhatsappChatBot(
              settings: config,
            )
          ],
        ));
  }
}
